File: backend/app/api/categories.py
"""app/api/categories.py

FastAPI router for user category endpoints.
Handles user category CRUD operations and management.
"""

from typing import List

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.api.auth import get_current_user
from app.database.db import get_session
from app.dtos.user_category_dtos import (
    UserCategoryBulkUpdateDTO,
    UserCategoryCreateDTO,
    UserCategoryReorderDTO,
    UserCategoryResponseDTO,
    UserCategoryUpdateDTO,
)
from app.models.user import User
from app.services.user_category_service import (
    BuiltInCategoryError,
    CategoryLimitExceededError,
    DuplicateUserCategoryError,
    UserCategoryNotFoundError,
    UserCategorySaveError,
    UserCategoryService,
)

router = APIRouter()


@router.get("", response_model=List[UserCategoryResponseDTO])
def list_categories(
    include_disabled: bool = Query(False, description="Include disabled categories"),
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """
    List all categories for the current user.

    Returns categories ordered by position.
    On first access, built-in categories are automatically seeded.

    Query params:
    - include_disabled: If true, includes disabled categories (default: false)
    """
    service = UserCategoryService(session, current_user.id)
    return service.get_all_categories(include_disabled=include_disabled)


@router.get("/{category_id}", response_model=UserCategoryResponseDTO)
def get_category(
    category_id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Get a single category by ID."""
    service = UserCategoryService(session, current_user.id)
    try:
        return service.get_category_by_id(category_id)
    except UserCategoryNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("", response_model=UserCategoryResponseDTO, status_code=201)
def create_category(
    category_data: UserCategoryCreateDTO,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """
    Create a new custom category.

    - label: Required, 1-50 characters
    - A URL-safe slug is automatically generated from the label
    - Maximum 50 custom categories per user
    """
    service = UserCategoryService(session, current_user.id)
    try:
        return service.create_category(category_data)
    except CategoryLimitExceededError as e:
        raise HTTPException(status_code=409, detail=str(e))
    except DuplicateUserCategoryError as e:
        raise HTTPException(status_code=409, detail=str(e))
    except UserCategorySaveError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.put("/{category_id}", response_model=UserCategoryResponseDTO)
def update_category(
    category_id: int,
    update_data: UserCategoryUpdateDTO,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """
    Update a category.

    - label: Optional, 1-50 characters (only for custom categories)
    - is_enabled: Optional, toggle category visibility

    Note: Built-in categories can only be enabled/disabled, not renamed.
    """
    service = UserCategoryService(session, current_user.id)
    try:
        return service.update_category(category_id, update_data)
    except UserCategoryNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except BuiltInCategoryError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except UserCategorySaveError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/{category_id}")
def delete_category(
    category_id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """
    Delete a custom category.

    Note: Built-in categories cannot be deleted, only disabled.
    """
    service = UserCategoryService(session, current_user.id)
    try:
        service.delete_category(category_id)
        return {"message": "Category deleted successfully"}
    except UserCategoryNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except BuiltInCategoryError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except UserCategorySaveError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.put("/reorder", response_model=List[UserCategoryResponseDTO])
def reorder_categories(
    reorder_data: UserCategoryReorderDTO,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """
    Reorder categories.

    - ordered_ids: List of category IDs in the desired order
    """
    service = UserCategoryService(session, current_user.id)
    try:
        return service.reorder_categories(reorder_data)
    except UserCategorySaveError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.put("/bulk", response_model=List[UserCategoryResponseDTO])
def bulk_update_categories(
    bulk_data: UserCategoryBulkUpdateDTO,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """
    Bulk update multiple categories.

    - categories: List of {id, is_enabled, position} objects
    """
    service = UserCategoryService(session, current_user.id)
    try:
        return service.bulk_update(bulk_data)
    except UserCategorySaveError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/reset", response_model=List[UserCategoryResponseDTO])
def reset_categories(
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """
    Reset categories to defaults.

    - Re-enables all built-in categories
    - Disables (but keeps) all custom categories
    - Restores default ordering (built-ins first, then custom)
    """
    service = UserCategoryService(session, current_user.id)
    try:
        return service.reset_to_defaults()
    except UserCategorySaveError as e:
        raise HTTPException(status_code=500, detail=str(e))

File: frontend/src/lib/api/categories.ts
import type {
  UserCategoryDTO,
  UserCategoryCreateDTO,
  UserCategoryUpdateDTO,
  UserCategoryReorderDTO,
  UserCategoryBulkUpdateDTO,
} from "@/types/category";
import { fetchApi } from "./base";

export const categoryApi = {
  /**
   * List all categories for the current user
   * @param token - Auth token for authenticated requests
   * @param includeDisabled - Include disabled categories (default: false)
   * @returns Array of user categories ordered by position
   */
  list: (
    token: string | null,
    includeDisabled = false
  ): Promise<UserCategoryDTO[]> =>
    fetchApi<UserCategoryDTO[]>(
      `/api/categories${includeDisabled ? "?include_disabled=true" : ""}`,
      undefined,
      token
    ),

  /**
   * Get a single category by ID
   * @param id - Category ID
   * @param token - Auth token for authenticated requests
   * @returns Category details
   */
  get: (id: number, token: string | null): Promise<UserCategoryDTO> =>
    fetchApi<UserCategoryDTO>(`/api/categories/${id}`, undefined, token),

  /**
   * Create a new custom category
   * @param data - Category data (label)
   * @param token - Auth token for authenticated requests
   * @returns Created category
   */
  create: (
    data: UserCategoryCreateDTO,
    token: string | null
  ): Promise<UserCategoryDTO> =>
    fetchApi<UserCategoryDTO>(
      "/api/categories",
      {
        method: "POST",
        body: JSON.stringify(data),
      },
      token
    ),

  /**
   * Update a category
   * @param id - Category ID
   * @param data - Updated category data
   * @param token - Auth token for authenticated requests
   * @returns Updated category
   */
  update: (
    id: number,
    data: UserCategoryUpdateDTO,
    token: string | null
  ): Promise<UserCategoryDTO> =>
    fetchApi<UserCategoryDTO>(
      `/api/categories/${id}`,
      {
        method: "PUT",
        body: JSON.stringify(data),
      },
      token
    ),

  /**
   * Delete a custom category
   * @param id - Category ID
   * @param token - Auth token for authenticated requests
   */
  delete: (id: number, token: string | null): Promise<void> =>
    fetchApi<void>(
      `/api/categories/${id}`,
      {
        method: "DELETE",
      },
      token
    ),

  /**
   * Reorder categories
   * @param data - Ordered list of category IDs
   * @param token - Auth token for authenticated requests
   * @returns All categories with updated positions
   */
  reorder: (
    data: UserCategoryReorderDTO,
    token: string | null
  ): Promise<UserCategoryDTO[]> =>
    fetchApi<UserCategoryDTO[]>(
      "/api/categories/reorder",
      {
        method: "PUT",
        body: JSON.stringify(data),
      },
      token
    ),

  /**
   * Bulk update multiple categories
   * @param data - List of category updates
   * @param token - Auth token for authenticated requests
   * @returns All categories with updated states
   */
  bulkUpdate: (
    data: UserCategoryBulkUpdateDTO,
    token: string | null
  ): Promise<UserCategoryDTO[]> =>
    fetchApi<UserCategoryDTO[]>(
      "/api/categories/bulk",
      {
        method: "PUT",
        body: JSON.stringify(data),
      },
      token
    ),

  /**
   * Reset categories to defaults
   * @param token - Auth token for authenticated requests
   * @returns All categories with reset states
   */
  reset: (token: string | null): Promise<UserCategoryDTO[]> =>
    fetchApi<UserCategoryDTO[]>(
      "/api/categories/reset",
      {
        method: "POST",
      },
      token
    ),
};
